custom/aaa/models.py

```diff
@@ -0,0 +1,149 @@
+"""Contains all models for the AAA Convergence Dashboard.
+
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
+      They are saved with the format [[date, value], [date, value]].
+      They are not guaranteed to be ordered by date.
+      They are not intended to be queried on, their interpretation should happen in python.
+      These history fields are saved in a separate model to ensure that the main model stays
+      fast and small for filtering.
+    * Each model contains all parts of the location hierarchy.
+    * Each model should support a future world where it is being accessed in a
+      sharded database (CitusDB).
+"""
+
+from __future__ import absolute_import, unicode_literals
+
+from django.contrib.postgres.fields import ArrayField, DateRangeField
+from django.db import models
+
+from corehq.apps.userreports.models import StaticDataSourceConfiguration, get_datasource_config
+from corehq.apps.userreports.util import get_table_name
+
+
+class LocationDenormalizedModel(models.Model):
+    """Abstract base model for containing all the location fields necessary for querying."""
+
+    # support multiple domains so QA is easier
+    domain = models.TextField(null=False)
+
+    # Shared location hierarchy
+    state_id = models.TextField(null=True)
+    district_id = models.TextField(null=True)
```

Comment:
I'm intentionally using denormalization for now. I'm attempting to balance getting something QA-able before the pilot (march 1) vs supporting future CitusDB backend vs supporting future reports vs similarities to the current CAS dashboard so hopefully we can consolidate the custom warehouses and its easier to spin up on reports.

I'm not sure if I have made the right trade-off but I think it was easiest to do this for now and possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs. It's possible that normalizing now with a foreign key to a location table would be more performant and easy to migrate, but with an impending Citus migration I'm less confident that I understand all the moving pieces there.

Reply:
> possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs

no data to back this up, but this sounds intuitively like a good option to me, and would be how a standard "star schema" dimensional data warehouse design would tackle the problem.

this is quite a lot of redundant storage assuming one of the low-level ids can deterministically generate the rest.

also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

Reply:
> also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

FWIW This doesn't happen right now, but it could in the future. Before this does happen on a large scale for CAS there should be more speccing and understanding of what they actually want. There's a case to be made that we would not want the districts state to change for historical data as it would point to the geography that was relevant at that time. 

There would be similar questions for beneficiaries moving between AWC or ASHAs. Should we report only the current ASHA, or have the beneficiary report the different ones? 

Not that any of that is actionable right now, but I've heard the talks of this workflow being added for the past year or so, but no progress has been made on actually doing it.